src/vs/editor/contrib/hover/browser/markdownHoverParticipant.ts

```diff
@@ -125,16 +170,148 @@ export class MarkdownHoverParticipant implements IEditorHoverParticipant<Markdow
 		}
 
 		const position = new Position(anchor.range.startLineNumber, anchor.range.startColumn);
-		return getHover(this._languageFeaturesService.hoverProvider, model, position, token)
+		return getHover(this._languageFeaturesService.hoverProvider, model, position, undefined, token)
 			.filter(item => !isEmptyMarkdownString(item.hover.contents))
 			.map(item => {
 				const rng = item.hover.range ? Range.lift(item.hover.range) : anchor.range;
-				return new MarkdownHover(this, rng, item.hover.contents, false, item.ordinal);
+				return new VerboseMarkdownHover(this, rng, item.hover.contents, false, item.ordinal, item.provider, item.hover.verbosityMetadata);
 			});
 	}
 
-	public renderHoverParts(context: IEditorHoverRenderContext, hoverParts: MarkdownHover[]): IDisposable {
-		return renderMarkdownHovers(context, hoverParts, this._editor, this._languageService, this._openerService);
+	public renderHoverParts(context: IEditorHoverRenderContext, hoverParts: VerboseMarkdownHover[]): IDisposable {
+		this._context = context;
+		this._verbosityLevels.clear();
+		this._providers = hoverParts.map(hoverPart => hoverPart.provider);
+		hoverParts.sort((a, b) => a.ordinal - b.ordinal);
+		const disposables = new DisposableStore();
+		for (const [hoverIndex, hoverPart] of hoverParts.entries()) {
+			const renderedMarkdown = this._renderMarkdownHoversAndActions(
+				hoverPart.contents,
+				hoverIndex,
+				hoverPart.verbosityMetadata,
+				disposables
+			);
+			this._verbosityLevels.set(hoverIndex, 0);
+			context.fragment.appendChild(renderedMarkdown);
+		}
+		return disposables;
+	}
+
+	public async changeFocusedHoverVerbosityLevel(increaseVerbosity: boolean): Promise<void> {
+		if (
+			this._focusMetadata.index === undefined
+			|| this._focusMetadata.element === undefined
```

Comment:
Can it be that one is undefined and the other not?

Reply:
No longer needed after making this._focusInfo itself of type FocusInfo or undefined